src/Core/Services/Implementations/UserService.cs

```diff
@@ -356,6 +361,124 @@ public async Task<bool> VerifyTwoFactorEmailAsync(User user, string token)
                 "2faEmail:" + email, token);
         }
 
+        public async Task<CredentialCreateOptions> StartWebAuthnRegistrationAsync(User user)
+        {
+            await _u2fRepository.DeleteManyByUserIdAsync(user.Id);
+
+            var fidoUser = new Fido2User
+            {
+                DisplayName = user.Name,
+                Name = user.Email,
+                Id = user.Id.ToByteArray(),
+            };
+
+            var providers = user.GetTwoFactorProviders();
+            if (providers == null)
+            {
+                providers = new Dictionary<TwoFactorProviderType, TwoFactorProvider>();
+            }
+            var provider = user.GetTwoFactorProvider(TwoFactorProviderType.WebAuthn);
+            if (provider == null)
+            {
+                provider = new TwoFactorProvider
+                {
+                    Enabled = false
+                };
+            }
+            if (provider.MetaData == null)
+            {
+                provider.MetaData = new Dictionary<string, object>();
+            }
+
+            var excludeCredentials = provider.MetaData
+                .Where(k => k.Key.StartsWith("Key"))
+                .Select(k => new TwoFactorProvider.WebAuthnData((dynamic)k.Value).Descriptor)
+                .ToList();
+
+            var options = _fido2.RequestNewCredential(fidoUser, excludeCredentials, AuthenticatorSelection.Default, AttestationConveyancePreference.Direct);
+
+            provider.MetaData["pending"] = new TwoFactorProvider.WebAuthnData{ Options = options.ToJson() };
+
+
+            if (providers.ContainsKey(TwoFactorProviderType.WebAuthn))
+            {
+                providers.Remove(TwoFactorProviderType.WebAuthn);
+            }
+
+            providers.Add(TwoFactorProviderType.WebAuthn, provider);
+            user.SetTwoFactorProviders(providers);
+            await UpdateTwoFactorProviderAsync(user, TwoFactorProviderType.WebAuthn, false);
```

Comment:
I am a little confused about why you call both of these. Doesn't `UpdateTwoFactorProviderAsync` call `user.SetTwoFactorProviders`?

Reply:
This is how it's done in other places. `UpdateTwoFactorProviderAsync` only receives the `user` as input, so I don't see a way for it to call `SetTwoFactorProviders`. But I might be missing something.

Reply:
I looked into it again, while `UpdateTwoFactorProviderAsync` calls `user.SetTwoFactorProviders` it only for setting `enabled`. Since we need to modify the data directly and not only change `enabled`, we need to call `user.SetTwoFactorProviders`.